tests/Integration/PrestaShopBundle/Controller/Sell/Catalog/MultiShopProductControllerTest.php

```diff
@@ -518,6 +505,37 @@ protected function getGridSelector(): string
         return '#product_grid_table';
     }
 
+    /**
+     * This is not ideal but since we now have a request listener that forces a proper initialization of the context
```

Comment:
I was thinking about one usecase: we do have sometimes legacy code that modifies the context at runtime, after it has been initialized.

Two real life examples:
- in backoffice "create an order page" I think we mess with the context Cart object
- I think there are pages where you can modify the context currency at runtime

Maybe we should anticipate these usecases and allow runtime re-construction of the context?

One possible solution:
- context is built at loading time on kernel events
- there are services ; like `ContextRebuilder` that can be called at runtime and re-create the Context from scratch using updated data

Reply:
Also, Context is modified back and forth when rendering PDFs because there is no way to switch language for rendering.

Reply:
You're right @matks but the use case you're mentioning is more the exception than the regular use case The refacto focuses on the global cases for now, we'll then have to replace all the usages of the legacy context pice by piece and when we'll reach the Order edition we'll think about how we should handle this properly

But we shouldn't make it simple to change the context, if we do we'll meet the exact same problems we already have with the current unstable context

@Hlavtox about your example it's quite a big problem, this indicates that the PDF rendering is wrongly designed if it requires this context switching, this is the kind of use cases that make the legacy context quite unstable Maybe we should refacto the PDF rendering properly instead of bending the new context refacto to match some unwanted behaviour